ssl/ech/ech_internal.c

```diff
@@ -160,4 +214,1260 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+        }
+    }
+    if (!WPACKET_close(&inner)  /* close the exts sub packet */
+        || !WPACKET_close(&inner) /* close the inner CH */
+        || !WPACKET_get_length(&inner, &innerlen)) { /* len for inner CH */
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    innerch_full = OPENSSL_malloc(innerlen);
+    if (innerch_full == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* Finally ditch the type and 3-octet length */
+    memcpy(innerch_full, inner_mem->data + 4, innerlen - 4);
+    OPENSSL_free(s->ext.ech.encoded_innerch);
+    s->ext.ech.encoded_innerch = innerch_full;
+    s->ext.ech.encoded_innerch_len = innerlen - 4;
+    /* and clean up */
+    rv = 1;
+err:
+    WPACKET_cleanup(&inner);
+    BUF_MEM_free(inner_mem);
+    return rv;
+}
+
+/*
+ * Find ECH acceptance signal in a SH
+ * hrr is 1 if this is for an HRR, otherwise for SH
+ * acbuf is (a preallocated) 8 octet buffer
+ * shbuf is a pointer to the SH buffer (incl. the type+3-octet length)
+ * shlen is the length of the SH buf
+ * return: 1 for success, 0 otherwise
+ */
+int ossl_ech_find_confirm(SSL_CONNECTION *s, int hrr,
+                          unsigned char acbuf[OSSL_ECH_SIGNAL_LEN],
+                          const unsigned char *shbuf, const size_t shlen)
+{
+    PACKET pkt;
+    const unsigned char *acp = NULL, *pp_tmp;
+    unsigned int pi_tmp, etype, elen;
+    int done = 0;
+
+    if (hrr == 0) {
+        if (shlen < CLIENT_VERSION_LEN + SSL3_RANDOM_SIZE)
+            return 0;
+        acp = shbuf + CLIENT_VERSION_LEN + SSL3_RANDOM_SIZE
+            - OSSL_ECH_SIGNAL_LEN;
+        memcpy(acbuf, acp, OSSL_ECH_SIGNAL_LEN);
+        return 1;
+    } else {
+        if (!PACKET_buf_init(&pkt, shbuf, shlen)
+            || !PACKET_get_net_2(&pkt, &pi_tmp)
+            || !PACKET_get_bytes(&pkt, &pp_tmp, SSL3_RANDOM_SIZE)
+            || !PACKET_get_1(&pkt, &pi_tmp) /* sessid len */
+            || !PACKET_get_bytes(&pkt, &pp_tmp, pi_tmp) /* sessid */
+            || !PACKET_get_net_2(&pkt, &pi_tmp) /* ciphersuite */
+            || !PACKET_get_1(&pkt, &pi_tmp) /* compression */
+            || !PACKET_get_net_2(&pkt, &pi_tmp)) /* len(extensions) */
+            return 0;
+        while (PACKET_remaining(&pkt) > 0 && done < 1) {
+            if (!PACKET_get_net_2(&pkt, &etype)
+                || !PACKET_get_net_2(&pkt, &elen))
+                return 0;
+            if (etype == TLSEXT_TYPE_ech) {
+                if (elen != OSSL_ECH_SIGNAL_LEN
+                    || !PACKET_get_bytes(&pkt, &acp, elen))
+                    return 0;
+                memcpy(acbuf, acp, elen);
+                done++;
+            } else {
+                if (!PACKET_get_bytes(&pkt, &pp_tmp, elen))
+                    return 0;
+            }
+        }
```

Comment:
that's a good question, looking at it now...

Reply:
so the reason is that, in the case of HRR, the ECH acceptance check is done before we start to process extensions (so we can swap from outer to inner handling on client) - I'm not sure what'd break if we tried defer the check/swap 'till we get into `tls_parse_stoc_ech`

Reply:
I'm not sure why anything would break particularly.

We simply call `tls_collect_extensions` prior to accepting ECH, followed by:
`tls_parse_extension(s, TLSEXT_IDX_ech, SSL_EXT_TLS1_3_HELLO_RETRY_REQUEST, extensions, NULL, 0))`

This should cause `tls_parse_stoc_ech` to be called.

Presumably if we then swap from outer to inner handling, we just have to call `tls_collect_extensions` again and ditch the previous `extensions` that we parsed?